Etl2575 - Error handling in camus: publish error

zliu41 opened the pull request:
- Retry if publish count to kafka fails.
- Fail job if max retires reached.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/CamusJob.java:
```diff
@@ -476,9 +478,23 @@ public void sendTrackingCounts(JobContext job, FileSystem fs, Path newExecutionO
-        finalCounts.postTrackingCountToKafka(job.getConfiguration(), props.getProperty(KAFKA_MONITOR_TIER), brokerList);
+        for (int i = 0; i <= publishCountMaxRetries; i++) {
+          try {
+            finalCounts.postTrackingCountToKafka(job.getConfiguration(), props.getProperty(KAFKA_MONITOR_TIER),
```
I think we need to put the retry inside the kafka publishing, retry on the high level loop would cause lots of duplicate publishing.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/CamusJob.java:
```diff
@@ -86,6 +86,8 @@
   public static final String ETL_EXECUTION_HISTORY_MAX_OF_QUOTA = "etl.execution.history.max.of.quota";
   public static final String ETL_FAIL_ON_ERRORS = "etl.fail.on.errors";
+  public static final String ETL_PUBLISH_COUNT_MAX_RETRIES = "etl.publish.count.max.retries";
+  public static final String ETL_PUBLISH_COUNT_MAX_RETRIES_DEFAULT = String.valueOf(Integer.MAX_VALUE);
```
The default probably should be 3.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/CamusJob.java:
```diff
@@ -476,9 +478,14 @@ public void sendTrackingCounts(JobContext job, FileSystem fs, Path newExecutionO
       String brokerList = getKafkaBrokers(job);
       for (EtlCounts finalCounts : allCounts.values()) {
-        finalCounts.postTrackingCountToKafka(job.getConfiguration(), props.getProperty(KAFKA_MONITOR_TIER), brokerList);
+        for (int i = 0; i <= publishCountMaxRetries; i++) {
```
Why do you still have that loop if the retry is handled inside ETLCounts already?

zliu41 replied:
Fixed.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/common/EtlCounts.java:
```diff
@@ -241,18 +241,24 @@ private void produceCount(String brokerList, ArrayList<byte[]> monitorSet) {
+            KeyedMessage keyedMessage = new KeyedMessage("TrackingMonitoringEvent", message);
+            producer.send(keyedMessage);
+            break;
+          } catch (RuntimeException e) {
```
It's not necessarily a RuntimeException

zliu41 replied:
Fixed.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/common/EtlCounts.java:
```diff
@@ -241,18 +241,24 @@ private void produceCount(String brokerList, ArrayList<byte[]> monitorSet) {
+          } catch (RuntimeException e) {
+            log.error("Publishing count for topic " + topic + " to " + brokerList.toString() + " has failed " + (i + 1)
+                + " times. " + (publishCountMaxRetries - i) + " more attempts will be made.");
+            if (i == publishCountMaxRetries) {
```
Add some sleep time.

zliu41 replied:
Fixed.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/common/EtlCounts.java:
```diff
@@ -241,18 +241,30 @@ private void produceCount(String brokerList, ArrayList<byte[]> monitorSet) {
+            log.error("Publishing count for topic " + topic + " to " + brokerList.toString() + " has failed " + (i + 1)
+                + " times. " + (publishCountMaxRetries - i) + " more attempts will be made.");
+            if (i == publishCountMaxRetries) {
+              throw new RuntimeException(e);
```
You might want to provide a error message (e.g. "max tried three times on publishing ...") together with the cause exception.

zliu41 replied:
Fixed.

zliu41 commented:
Added unit test for publishing count failure.
- Added a flag (useMockProducerForUnitTest) in EtlCounts.java. If flag==true, it will use a mocked Producer, which throws Exception in producer.send().
- Added 3 dummy classes (DummyEncoder, DummyGenericRecord, DummyMonitoringEvent). These classes are instantiated using reflection, so cannot be mocked.

zliu41 commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/CamusJob.java:
```diff
@@ -356,7 +362,9 @@ public int compare(FileStatus f1, FileStatus f2) {
     // Send Tracking counts to Kafka
-    sendTrackingCounts(job, fs, newExecutionOutput);
+    String etlCountsClassName = props.getProperty(ETL_COUNTS_CLASS, ETL_COUNTS_CLASS_DEFAULT);
+    Class<? extends EtlCounts> etlCountsClass = (Class<? extends EtlCounts>) Class.forName(etlCountsClassName);
```
The purpose of this is to be able to use class EtlCountsForUnitTest, instead of EtlCounts, in unit test.

HenryCaiHaiying replied:
OK

zliu41 commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/CamusJob.java:
```diff
@@ -448,10 +458,10 @@ public void sendTrackingCounts(JobContext job, FileSystem fs, Path newExecutionO
         mapper.configure(DeserializationConfig.Feature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+
         ArrayList<EtlCounts> countsObjects =
             mapper.readValue(buffer.toString(), new TypeReference<ArrayList<EtlCounts>>() {
```
Ideally I would like to pass `new TypeReference<ArrayList<EtlCountsForUnitTest>>()` in unit test. However, this doesn't seem to be easily doable. So I still pass `new TypeReference<ArrayList<EtlCounts>>()` here, and then in line 503, convert `EtlCounts finalEtlCounts` to `etlCountsClass`, which is `EtlCounts` normally, and `EtlCountsForUnitTest` in unit test.

zliu41 commented on camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/CamusJobTestWithMock.java:
```diff
@@ -308,6 +311,70 @@ private OffsetResponse mockOffsetResponseThirdTrySucceed(List<MyMessage> myMessa
   }
 
+  @Test(expected = RuntimeException.class)
+  public void testJobFailDueToPublishCountsException() throws Exception {
```
negative test

zliu41 commented on camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/CamusJobTestWithMock.java:
```diff
@@ -308,6 +311,70 @@ private OffsetResponse mockOffsetResponseThirdTrySucceed(List<MyMessage> myMessa
+  }
+
+  @Test
+  public void testJobPublishCountsThirdTrySucceed() throws Exception {
```
positive test: succeed in the last try.

zliu41 commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/common/EtlCounts.java:
```diff
@@ -45,7 +46,7 @@
   private long lastTimestamp;
   private long firstTimestamp = Long.MAX_VALUE;
-  private HashMap<String, Source> counts;
+  protected HashMap<String, Source> counts;
```
changed to protected because it needs to be accessed in subclass EtlCountsForUnitTest in unit test.

HenryCaiHaiying replied:
OK.

zliu41 commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/common/EtlCounts.java:
```diff
@@ -65,6 +66,11 @@ public EtlCounts(String topic, long granularity) {
     this(topic, granularity, System.currentTimeMillis());
   }
 
+  public EtlCounts(EtlCounts other) {
```
Added copy constructor for the purpose of converting EtlCounts object to either EtlCounts object or EtlCountsForUnitTest object (line 503 in CamusJob.java).

zliu41 commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/common/EtlCounts.java:
```diff
@@ -229,6 +235,10 @@ public void postTrackingCountToKafka(Configuration conf, String tier, String bro
 
+  protected String getMonitoringEventClass(Configuration conf) {
+    return conf.get(MONITORING_EVENT_CLASS);
+  }
```
put `conf.get(MONITORING_EVENT_CLASS)` as a method, so that it can be overridden in subclass EtlCountsForUnitTest.

HenryCaiHaiying replied:
OK.

zliu41 commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/common/EtlCounts.java:
```diff
@@ -238,21 +248,40 @@ private void produceCount(String brokerList, ArrayList<byte[]> monitorSet) {
 
+  protected Producer createProducer(Properties props) {
+    return new Producer(new ProducerConfig(props));
+  }
```
put `new Producer(new ProducerConfig(props))` as a method, so that it can be overridden in subclass EtlCountsForUnitTest.

HenryCaiHaiying replied:
OK.

HenryCaiHaiying commented:
Looks good to be merged.

zliu41 commented:
- Modified unit test based on the new mock framework.
- mvn clean package passed
- regression test on nertz passed.
